FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m

```diff
@@ -173,31 +208,56 @@ - (nullable NSData *)randomChallengeFromResponseBody:(NSData *)response error:(N
     return rejectedPromise;
   }
 
-  return [FBLPromise
-      onQueue:[self backgroundQueue]
-           do:^id _Nullable {
-             NSError *encodingError;
-             NSData *payloadJSON = [NSJSONSerialization dataWithJSONObject:@{
-               kRequestFieldKeyID : keyID,
-               kRequestFieldAttestation : [self base64StringWithData:attestation],
-               kRequestFieldChallenge : [self base64StringWithData:challenge]
-             }
-                                                                   options:0
-                                                                     error:&encodingError];
-
-             if (payloadJSON != nil) {
-               return payloadJSON;
-             } else {
-               return [FIRAppCheckErrorUtil JSONSerializationError:encodingError];
-             }
-           }];
+  return [FBLPromise onQueue:[self backgroundQueue]
+                          do:^id {
+                            id JSONObject = @{
+                              kRequestFieldKeyID : keyID,
+                              kRequestFieldAttestation : [self base64StringWithData:attestation],
+                              kRequestFieldChallenge : [self base64StringWithData:challenge]
+                            };
+
+                            return [self HTTPBodyWithJSONObject:JSONObject];
+                          }];
+}
+
+- (FBLPromise<NSData *> *)HTTPBodyWithJSONObject:(nonnull id)JSONObject {
+  NSError *encodingError;
+  NSData *payloadJSON = [NSJSONSerialization dataWithJSONObject:JSONObject
+                                                        options:0
+                                                          error:&encodingError];
+  __auto_type HTTPBodyPromise = [FBLPromise pendingPromise];
```

Comment:
nit: I think the inferred type for `HTTPBodyPromise` here will be as generic as just `FBLPromise *` which basically means not event a suggestion on a possibly invalid type passed in `fulfill:` method. For the cases when the right-hand statement doesn't provide enough information on the types it's better to use explicit type declaration (like `FBLPromise<NSData *> *`) instead of `__auto_type`.

Reply:
Thanks for clarifying on this!!